Ignore DefaultCredentials when doing Basic & Digest auth in SocketsHttpHandler

makazeu opened the pull request:
Fixes #113145.

As described in https://github.com/dotnet/runtime/issues/113145#issuecomment-2710357021

makazeu committed:
Ignore DefaultCredentials in Basic Auth

makazeu commented:
cc @MihaZupan

makazeu committed:
Fix formatting in AuthenticationHelper for Basic Auth credential check

makazeu committed:
Remove unnecessary blank line in AuthenticationHelper

MihaZupan commented:
Can you please add a test that validates the new behavior?
Default credentials are set, 401 response, only 1 request is made

makazeu committed:
Add SocketsHttpHandler_UseDefaultCredentials_OneRequestOnlyWhen401 test

makazeu committed:
Add unit test

Add unit test

makazeu committed:
update

makazeu committed:
update

makazeu committed:
update

makazeu committed:
update

wfurt commented on src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.cs:
```diff
@@ -266,6 +266,12 @@ await TrySetDigestAuthToken(request, challenge.Credential, digestResponse, isPro
                         break;
 
                     case AuthenticationType.Basic:
+                        if (CredentialCache.DefaultCredentials == credentials)
```
Should we duplicate this for AuthenticationType.Digest as well? It would have same problem, right?

makazeu committed:
Ignore DefaultCredentials in Digest auth

makazeu committed:
Update UT

makazeu committed:
Rename test method for clarity on Basic Authentication behavior

makazeu committed:
update

makazeu committed:
update

MihaZupan commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -1299,9 +1299,44 @@ public sealed class SocketsHttpHandlerTest_AutoRedirect : HttpClientHandlerTest_
+        [Fact]
+        public async Task SocketsHttpHandler_UseDefaultCredentials_OneRequestOnlyForBasicAuth()
+        {
+            var requestCount = 0;
+            await LoopbackServerFactory.CreateClientAndServerAsync(
+                async url =>
+                {
+                    using (var handler = new SocketsHttpHandler())
+                    using (var invoker = new HttpMessageInvoker(handler))
+                    {
+                        handler.Credentials = CredentialCache.DefaultCredentials;
+                        var request = new HttpRequestMessage(HttpMethod.Get, url);
+                        var response = await invoker.SendAsync(request, CancellationToken.None);
+                        Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
+                    }
+                },
+                async server =>
+                {
+                    var responseHeader = new[] { new HttpHeaderData("WWW-Authenticate", "Basic realm=\"Test Realm\"") };
+                    await server.HandleRequestAsync(HttpStatusCode.Unauthorized, responseHeader);
+                    requestCount++;
+
+                    // Only one request should be sent.
+                    await IgnoreExceptions(async () => 
+                    {
+                        await server.HandleRequestAsync(HttpStatusCode.Unauthorized, responseHeader).WaitAsync(TimeSpan.FromSeconds(10));
+                        requestCount++;
+                    });
+                }
+            );
+
+            Assert.Equal(1, requestCount);
```
```suggestion
            await LoopbackServerFactory.CreateClientAndServerAsync(
                async url =>
                {
                    using var handler = new SocketsHttpHandler();
                    using var client = new HttpClient(handler);
                    handler.Credentials = CredentialCache.DefaultCredentials;

                    using HttpResponseMessage response = await invoker.GetAsync(url);
                    Assert.Equal(HttpStatusCode.Unauthorized, response.StatusCode);
                },
                async server =>
                {
                    var responseHeader = new[] { new HttpHeaderData("WWW-Authenticate", "Basic realm=\"Test Realm\"") };
                    await server.HandleRequestAsync(HttpStatusCode.Unauthorized, responseHeader);
                }
            );
```

It's okay to validate that a single request was sent just by the fact that the request succeeded here.
We don't want to make the test take 10 seconds every time in order to hit a timeout.

MihaZupan commented on src/libraries/System.Net.Http/tests/FunctionalTests/SocketsHttpHandlerTest.cs:
```diff
@@ -1299,9 +1299,44 @@ public sealed class SocketsHttpHandlerTest_AutoRedirect : HttpClientHandlerTest_
     {
         public SocketsHttpHandler_DefaultCredentialsTest(ITestOutputHelper output) : base(output) { }
+
+        [Fact]
```
Please make this a theory to cover both `Basic` and `Digest`.

MihaZupan reviewed (approved):
Thank you

makazeu committed:
Merge branch 'main' into ignore-defaultcredentials-in-basicauth

makazeu committed:
Refactor SocketsHttpHandler test to use Theory for authentication types

Updated the SocketsHttpHandler_UseDefaultCredentials test to utilize [Theory] and [InlineData] attributes, allowing for testing with both Basic and Digest authentication types.

makazeu committed:
Merge branch 'main' into ignore-defaultcredentials-in-basicauth

wfurt reviewed (approved):
LGTM. thans @makazeu
